server/executor/transaction_runner.go

```diff
@@ -0,0 +1,160 @@
+	Run(context.Context, model.Transaction, model.RunMetadata, model.Environment) model.TransactionRun
+}
+
+func NewTransactionRunner(runner Runner, db model.Repository, config config.Config) PersistentTransactionRunner {
+	return PersistentTransactionRunner{
+		testRunner:          runner,
+		db:                  db,
+		executionChannel:    make(chan transactionRunJob, 1),
+		exit:                make(chan bool),
+		checkTestStateDelay: config.PoolingRetryDelay() / 2,
+	}
+}
+
+type transactionRunJob struct {
+	ctx context.Context
+	run model.TransactionRun
+}
+
+type PersistentTransactionRunner struct {
+	testRunner          Runner
+	db                  model.Repository
+	checkTestStateDelay time.Duration
+	executionChannel    chan transactionRunJob
+	exit                chan bool
+}
+
+func (r PersistentTransactionRunner) Run(ctx context.Context, transaction model.Transaction, metadata model.RunMetadata, environment model.Environment) model.TransactionRun {
+	run := model.NewTransactionRun(transaction)
+	run.Metadata = metadata
+	run.Environment = environment
+
+	run, _ = r.db.CreateTransactionRun(ctx, run)
+
+	r.executionChannel <- transactionRunJob{
+		ctx: ctx,
+		run: run,
+	}
+
+	return run
+}
+
+func (r PersistentTransactionRunner) Stop() {
+	r.exit <- true
+}
+
+func (r PersistentTransactionRunner) Start(workers int) {
+	for i := 0; i < workers; i++ {
+		go func() {
+			fmt.Println("PersistentTransactionRunner start goroutine")
+			for {
+				select {
+				case <-r.exit:
+					fmt.Println("PersistentTransactionRunner exit goroutine")
+					return
+				case job := <-r.executionChannel:
+					err := r.runTransaction(job.ctx, job.run)
+					if err != nil {
+						panic(err)
+					}
+				}
+			}
+		}()
+	}
+}
+
+func (r PersistentTransactionRunner) runTransaction(ctx context.Context, run model.TransactionRun) error {
+	run.State = model.TransactionRunStateExecuting
+	environment := run.Environment
+
+	var err error
+
+	for i := range run.Steps {
+		run, err = r.runTransactionStep(ctx, run, i, environment)
+		if err != nil {
+			return fmt.Errorf("could not execute step %d of transaction %s: %w", i, run.TransactionID, err)
+		}
+
+		if run.State == model.TransactionRunStateExecuting {
+			environment = r.patchEnvironment(environment, run)
+		}
+	}
+
+	if !run.State.IsFinal() {
+		run.State = model.TransactionRunStateFinished
+	}
+
+	return r.db.UpdateTransactionRun(ctx, run)
+}
+
+func (r PersistentTransactionRunner) runTransactionStep(ctx context.Context, transactionRun model.TransactionRun, stepIndex int, environment model.Environment) (model.TransactionRun, error) {
+	step := transactionRun.Steps[stepIndex]
+	testRun := r.testRunner.Run(ctx, step, transactionRun.Metadata, environment)
+
+	testRun, err := r.waitTestRunIsFinished(ctx, testRun)
+	if err != nil {
+		return model.TransactionRun{}, fmt.Errorf("could not wait for step execution: %w", err)
+	}
+
+	transactionRun.StepRuns = append(transactionRun.StepRuns, testRun)
```

Comment:
runs tend to be really big, since they hold the trace. Is it really necessary to have them in memory like this? you could hold a `map[testRunID]status` or something similar holding a small subset of the info if it's required somewhere else (which I don't think it is, except for transmiting it to the FE)